src/Futhark/CodeGen/ImpGen/GPU/SegRed.hs

```diff
@@ -105,48 +143,132 @@ compileSegRed' ::
-          use_small_segments = segment_size * 2 .<. group_size'
+      let segment_size = pe64 $ last $ segSpaceDims space
+          use_small_segments = segment_size * 2 .<. group_size_E * chunk_E
       sIf
         use_small_segments
-        (smallSegmentsReduction pat num_groups group_size space reds body)
-        (largeSegmentsReduction pat num_groups group_size space reds body)
+        (compileReduction smallSegmentsReduction)
+        (compileReduction largeSegmentsReduction)
   where
+    compileReduction f =
+      f segred_kind pat num_groups group_size chunk space segbinops map_body_cont
+
+    chunk
+      | NoncommPrimSegred <- segred_kind =
+          intConst Int64 $ getChunkSize param_types
+      | otherwise = intConst Int64 1
+
+    segred_kind
+      | mconcat (map segBinOpComm segbinops) == Noncommutative,
+        all primType param_types =
+          NoncommPrimSegred
+      | otherwise = GeneralSegred
+
+    param_types = map paramType $ concatMap paramOf segbinops
+
     num_groups = gridNumGroups grid
     group_size = gridGroupSize grid
+    group_size_E = pe64 $ unCount group_size
+    chunk_E = pe64 chunk
 
 -- | Prepare intermediate arrays for the reduction.  Prim-typed
 -- arguments go in local memory (so we need to do the allocation of
 -- those arrays inside the kernel), while array-typed arguments go in
--- global memory.  Allocations for the former have already been
+-- global memory.  Allocations for the latter have already been
 -- performed.  This policy is baked into how the allocations are done
 -- in ExplicitAllocations.
-intermediateArrays ::
+--
+-- For more info about the intermediate arrays used for the different reduction
+-- kernels, see note [IntermArrays].
+makeIntermArrays ::
+  SegredKind ->
   Imp.TExp Int64 ->
-  Count GroupSize SubExp ->
-  SegBinOp GPUMem ->
-  InKernelGen [VName]
-intermediateArrays group_id (Count group_size) (SegBinOp _ red_op nes _) = do
-  let red_op_params = lambdaParams red_op
-      (red_acc_params, _) = splitAt (length nes) red_op_params
-  forM red_acc_params $ \p ->
-    case paramDec p of
-      MemArray pt shape _ (ArrayIn mem _) -> do
-        let shape' = Shape [group_size] <> shape
-        sArray "red_arr" pt shape' mem $
-          LMAD.iota (group_id * product (map pe64 (shapeDims shape'))) (map pe64 $ shapeDims shape')
-      _ -> do
-        let pt = elemType $ paramType p
-            shape = Shape [group_size]
-        sAllocArray "red_arr" pt shape $ Space "local"
+  SubExp ->
+  SubExp ->
+  [[Param LParamMem]] ->
+  InKernelGen [SegRedIntermediateArrays]
+makeIntermArrays NoncommPrimSegred _ group_size chunk params = do
+  group_worksize <- tvSize <$> dPrimV "group_worksize" group_worksize_E
+
+  -- compute total amount of lmem needed for the group reduction arrays as well
+  -- as the offset into the total pool of lmem of each such array.
+
+  -- TODO: we essentially compute `group_reds_lmem_requirement` twice, since it
```

Comment:
Is this actually a TODO?  What is there to do?

Reply:
Sorry, yes, this should not have been a TODO but rather a question out of curiosity. Never mind this.